File: rximagepicker_support_wechat/src/main/java/com/qingmei2/rximagepicker_extension_wechat/ui/widget/WechatCheckView.java
package com.qingmei2.rximagepicker_extension_wechat.ui.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.RectF;
import android.os.Build;
import android.util.AttributeSet;

import com.qingmei2.rximagepicker_extension.ui.widget.CheckView;

public class WechatCheckView extends CheckView {

    private static final float WECHAT_STROKE_WIDTH = 1.5f; // dp
    private static final float WECHAT_STROKE_CONNER = 5f; // dp

    private RectF rect;
    private Rect drawableRect;

    private static final int PROOFREAD_SIZE = 8;

    public WechatCheckView(Context context) {
        super(context);
    }

    public WechatCheckView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public WechatCheckView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void init(Context context) {
        super.init(context);
        getMStrokePaint().setStrokeWidth(WECHAT_STROKE_WIDTH * getMDensity());

        // draw white stroke
        float dx = getMDensity() * Companion.getSIZE() / 4;
        float dy = getMDensity() * Companion.getSIZE() / 4;
        rect = new RectF(2 * dx - PROOFREAD_SIZE, dy - PROOFREAD_SIZE,
                3 * dx + PROOFREAD_SIZE, 2 * dy + PROOFREAD_SIZE);
        drawableRect = new Rect((int) rect.left, (int) rect.top, (int) rect.right, (int) rect.bottom);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        // draw outer and inner shadow
        initShadowPaint();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            drawRoundRectCheckView(canvas, rect, getMStrokePaint());
        } else {
            drawRectCheckView(canvas, rect, getMStrokePaint());
        }

        // draw content
        if (getMCountable()) {
            if (getMCheckedNum() != Companion.getUNCHECKED()) {
                initBackgroundPaint();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    drawRoundRectCheckView(canvas, rect, getMBackgroundPaint());
                } else {
                    drawRectCheckView(canvas, rect, getMBackgroundPaint());
                }
                initTextPaint();
                String text = String.valueOf(getMCheckedNum());
                int baseX = (int) (rect.width() - getMTextPaint().measureText(text)) / 2;
                int baseY = (int) (rect.height() - getMTextPaint().descent() - getMTextPaint().ascent()) / 2;
                canvas.drawText(text, baseX + rect.left, baseY + rect.top, getMTextPaint());
            }
        } else {
            if (getMChecked()) {
                initBackgroundPaint();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    drawRoundRectCheckView(canvas, rect, getMBackgroundPaint());
                } else {
                    drawRectCheckView(canvas, rect, getMBackgroundPaint());
                }
                getMCheckDrawable().setBounds(drawableRect);
                getMCheckDrawable().draw(canvas);
            }
        }

        // enable hint
        setAlpha(getMEnabled() ? 1.0f : 0.5f);
    }

    private void drawRectCheckView(Canvas canvas, RectF rect, Paint paint) {
        canvas.drawRect(rect, paint);
    }

    @TargetApi(value = Build.VERSION_CODES.LOLLIPOP)
    private void drawRoundRectCheckView(Canvas canvas, RectF rect, Paint paint) {
        canvas.drawRoundRect(rect, WECHAT_STROKE_CONNER, WECHAT_STROKE_CONNER, paint);
    }
}

File: rximagepicker_support_wechat/src/main/java/com/qingmei2/rximagepicker_extension_wechat/ui/widget/WechatCheckView.kt
package com.qingmei2.rximagepicker_extension_wechat.ui.widget

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.RectF
import android.os.Build
import android.util.AttributeSet

import com.qingmei2.rximagepicker_extension.ui.widget.CheckView

open class WechatCheckView : CheckView {

    private var rect: RectF? = null
    private var drawableRect: Rect? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun init(context: Context) {
        super.init(context)
        mStrokePaint.strokeWidth = WECHAT_STROKE_WIDTH * mDensity

        // draw white stroke
        val dx = mDensity * SIZE / 4
        val dy = mDensity * SIZE / 4
        rect = RectF(2 * dx - PROOFREAD_SIZE, dy - PROOFREAD_SIZE,
                3 * dx + PROOFREAD_SIZE, 2 * dy + PROOFREAD_SIZE)
        drawableRect = Rect(rect!!.left.toInt(), rect!!.top.toInt(), rect!!.right.toInt(), rect!!.bottom.toInt())
    }

    override fun onDraw(canvas: Canvas) {
        // draw outer and inner shadow
        initShadowPaint()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            drawRoundRectCheckView(canvas, rect, mStrokePaint)
        } else {
            drawRectCheckView(canvas, rect, mStrokePaint)
        }

        // draw content
        if (mCountable) {
            if (mCheckedNum != CheckView.UNCHECKED) {
                initBackgroundPaint()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    drawRoundRectCheckView(canvas, rect, mBackgroundPaint)
                } else {
                    drawRectCheckView(canvas, rect, mBackgroundPaint)
                }
                initTextPaint()
                val text = mCheckedNum.toString()
                val baseX = (rect!!.width() - mTextPaint!!.measureText(text)).toInt() / 2
                val baseY = (rect!!.height() - mTextPaint!!.descent() - mTextPaint!!.ascent()).toInt() / 2
                canvas.drawText(text, baseX + rect!!.left, baseY + rect!!.top, mTextPaint!!)
            }
        } else {
            if (mChecked) {
                initBackgroundPaint()
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    drawRoundRectCheckView(canvas, rect, mBackgroundPaint)
                } else {
                    drawRectCheckView(canvas, rect, mBackgroundPaint)
                }
                mCheckDrawable!!.bounds = drawableRect!!
                mCheckDrawable!!.draw(canvas)
            }
        }

        // enable hint
        alpha = if (mEnabled) 1.0f else 0.5f
    }

    private fun drawRectCheckView(canvas: Canvas, rect: RectF?, paint: Paint?) {
        canvas.drawRect(rect!!, paint!!)
    }

    @TargetApi(value = Build.VERSION_CODES.LOLLIPOP)
    private fun drawRoundRectCheckView(canvas: Canvas, rect: RectF?, paint: Paint?) {
        canvas.drawRoundRect(rect!!, WECHAT_STROKE_CONNER, WECHAT_STROKE_CONNER, paint!!)
    }

    companion object {

        private const val WECHAT_STROKE_WIDTH = 1.5f // dp
        private const val WECHAT_STROKE_CONNER = 5f // dp
        private const val SIZE = 48 // dp

        private const val PROOFREAD_SIZE = 8
    }
}
